### PR TITLE
Initial contribution to KrafFeaturesCheck.

### LINKED ISSUE #58: Should every binding be mentioned in features\openhab-addons\src\main\feature\feature.xml
Since this is a file which is not changed, so no incorrect change but a missing change, its harder to detect in a PR. If it is required we could consider adding a rule for this?

### PR DESCRIPTION
**Pull request by svilenvul**
Closes #58

Signed-off-by: Svilen Valkanov <svilen.valkanov@musala.com>

### CODE REVIEW

File: src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        setFileExtensions(XML_EXTENSION);
+    }
+
+    public void setFeatureXmlPath(String featureXmlPath) {
+        this.featureXmlPath = featureXmlPath;
+    }
+
+    @Override
+    protected void processFiltered(File file, FileText fileText) throws CheckstyleException {
+        if (file.getName().equals(POM_XML_FILE_NAME)) {
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java]: Position literals first in `String` comparisons - that way if the `String` is `null` you won't get a `NullPointerException`, it'll just return `false`. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

File: src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            String bundleId = getBundleId(fileText);
+            if (bundleId == null) {
+                logger.error(this.getClass().getSimpleName()
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java]: Should you use placeholders here?

File: src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+            Path featurePath = resolveRecursively(file.toPath(), Paths.get(featureXmlPath));
+            if (featurePath == null) {
+                logger.error(
+                        this.getClass().getSimpleName() + " will be skipped. Could not find file " + featureXmlPath);
+                return;
+            }
+
+            try {
+                FileText featureFileText = new FileText(featurePath.toFile(), "UTF-8");
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java]: I would prefer to use `StandardCharsets.UTF_8.name()` [see](https://docs.oracle.com/javase/7/docs/api/java/nio/charset/StandardCharsets.html#UTF_8) 

And you can import the `UTF_8` to make it a little more compact.

File: src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    private String getBundleId(FileText fileText) throws CheckstyleException {
+        Document featureXML = parseDomDocumentFromFile(fileText);
+
+        Node artifactId = getFirstNode(featureXML, POM_ARTIFACT_ID_XPATH_EXPRESSION);
+        Node groupId = getFirstNode(featureXML, POM_GROUP_ID_XPATH_EXPRESSION);
+        Node parentGroupId = getFirstNode(featureXML, POM_PARENT_GROUP_ID_XPATH_EXPRESSION);
+
+        // Maven allows us to skip adding the group ID if the parent element has group ID
```
Comment (martinvw) [inline src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java]: 👍

File: src/test/java/org/openhab/tools/analysis/checkstyle/test/KarafFeatureCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    public void testIncludedBundleWithParentGroupIdOnly() throws Exception {
+        verify("includedBundleWithParentGroupIdOnly", ArrayUtils.EMPTY_STRING_ARRAY);
+    }
+
+    @Test
+    public void testInvalidBundle() throws Exception {
+        verify("invalidBundle", ArrayUtils.EMPTY_STRING_ARRAY);
+        // Error will be logged on the console
```
Comment (martinvw) [inline src/test/java/org/openhab/tools/analysis/checkstyle/test/KarafFeatureCheckTest.java]: There are actually ways to validate this by adding a log appender for the specific class and capturing the specific message.

File: src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java (hunk 2 above)
Comment (svilenvul) [reply src/main/java/org/openhab/tools/analysis/checkstyle/KarafFeatureCheck.java]: Unfortunately we are not using SLF4j anymore, org.apache.commons.logging does not support placeholders.

**Comment by martinvw** [comment]
Thanks

**Pull request merged**